Redistribute stateful material properties

roystgnr opened the pull request:
This should close #4532

## Reason
We previously could not handle any repartitioning in problems using stateful material properties; in particular this made adaptive mesh refinement less efficient in parallel.

## Design
We now attach a GhostingFunctor subclass, RedistributeProperties, which uses the libMesh redistribute() callback as an opportunity to serialize stored material property data and send it to whichever processor (or processors, when we're repartitioning in conjunction with refinement) will need it next.

## Impact
Two public `MaterialPropertyStorage` APIs have had a new parameter added.  I think these are only being used internally, but any external users would have to modify their code to match.

Three `MaterialPropertyStorage` setter APIs have had a name change; this helped with debugging, and again I think they're only being used internally, but we might revert this if there turn out to be apps trying to do tricky things under the framework hood.

The range returned by `refinedElementRange` now includes *all* parents of local just-refined children, not just local parents.  I think this is more consistent with the API comment, and it's a better fit for our internal use of that method, but if anyone else is using refinement with repartitioning and was assuming the former behavior then they'll have a bit of a surprise.

Hopefully, no further impact other than "problems with stateful material problems will now automatically allow repartitioning if the users don't manually disable it".  Debugging this code was quite horrific, so I do worry about the possibility of corner cases I missed, but if there still are any then I believe "manually disable repartitioning" would be an easy workaround while a further fix was worked on.

roystgnr committed:
Boilerplate RedistributeProperties GhostingFunctor

roystgnr committed:
Keep track of MaterialPropertyStorage we're handed

roystgnr committed:
MaterialPropertyStorage::PropsType typedef

This makes things a bit more readable

roystgnr committed:
Change misleading local function name

roystgnr committed:
RedistributeProperties needs initProps access

roystgnr committed:
Do initProps + deletions in RedistributeProperties

roystgnr committed:
assert, not segfault, on store/load size mismatch

This made debugging a bit easier

roystgnr committed:
Cache local refined elements' nonlocal parents too

roystgnr committed:
RedistributeProperties fixes for AMR, repeat use

Deleting redistributed properties avoids us accessing them later when
the elements they're tied to have been deleted too.  Redistribute
properties that are necessary for prolongation and restriction on
adapted meshes.

roystgnr committed:
There's no reason contains() shouldn't be const

roystgnr committed:
Get rid of non-const encapsulation-violators

Hopefully these aren't used in apps.

roystgnr committed:
Assertions in MaterialPropertyStorage accessors

We don't want to accidentally create a new entry, but operator[] will
happily do that without warning.

roystgnr committed:
More tightening MaterialPropertyStorage accessors

Splitting out the non-const accessors into their own functions makes it
a little easier to catch any bad writes.

roystgnr committed:
Don't try to prolong onto non-local children

Doing so leaves us with something in storage that we won't know what to
do with afterward.

roystgnr committed:
Don't add older-prop entries we aren't using

This is more efficient and it makes it easier to verify that the storage
is in a consistent state.

roystgnr committed:
Another assert

roystgnr committed:
Use a local spin_mutex in MaterialPropertyStorage

This should be more efficient when highly threaded.

roystgnr committed:
Use RedistributeProperties when needed

This is enough to solve #4532 for me.  With these changes we should no
longer be automatically using workarounds instead.  I'll fix up the
regression tests where we were manually using workarounds too.

roystgnr committed:
repartitioning+distribution in stateful_prop test

Works for me now!

roystgnr committed:
Per-map initProps

This is slightly cleaner code, but more importantly it'll let me work
safely in one map at a time in RedistributeProperties

roystgnr committed:
RedistributeProperties fixes

It seems to be passing whatever I throw at it now

roystgnr committed:
Redistribution in another stateful_prop test

I may have actually fixed all the bugs here now

roystgnr committed:
Workaround structured-binding+lambda error

We were getting hit by some C++ standards weirdness on some compilers
before - see https://stackoverflow.com/a/46115028

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1389053/) on c818022 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23603/site)

This comment will be updated on new commits.

lindsayad commented:
This might be a record for delta between PR and issue number

roystgnr committed:
Remove overzealous assertion

Maybe at some point I can clean up the design here a little more...

roystgnr committed:
Switch back to the catch-all Threads::spin_mtx

This can't be the right solution, but if it was working before then it's
probably the right workaround for now.

roystgnr commented:
Well, that's fun.  I can't replicate a problem when running threaded locally, and reverting my one threading-related change doesn't fix it.  This may end up sitting in limbo for a bit.

roystgnr committed:
HashMap::find(k) should be thread-safe too

I wonder if we should really be using public inheritance here; it seems
like every unordered_map method is a booby-trap waiting to happen when
running with threads.

roystgnr commented:
Alternatively, I might just replicate the problem, understand the problem, and fix the problem, right now.  Boo-yah.

If this passes CI (likely, IMHO, if "problem" really is singular and not plural) then it ought to be ready to review.  If it also passes review off-the-bat (less likely...) then it ought to be ready to merge.

roystgnr commented:
"problem" is plural.  I can replicate the dirackernels test failure, but won't have time to figure it out until I'm back.

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1395141/) on 8709669 : invalidated by @lindsayad

lindsayad commented:
Oh, you can reproduce ... so my invalidation will do no good. Boo

roystgnr committed:
Fix for DiracPointsWriter tests in parallel

Turning on repartitioning was what triggered a test failure here;
previously every element in the pls02_autopoints test was forced to be
on proc 0 even in parallel because that's where the original one coarse
element was.

roystgnr commented:
Assuming that test fix works in CI as well as it does on my workstation, this will be ready for review.

Kicking myself: I'd have fixed that test weeks ago, if only I'd predicted that the solution would be "fix a simple bug that was merely newly triggered in someone else's straightforward code" and not "fix another stupidly complicated corner case interaction with the new material property redistribution code".  In my defense, I'd just finished doing the latter so many times in a row that I didn't even think to check the former.

roystgnr commented:
That python failure is definitely not on me ... but it also looks like an actual compatibility issue, not anything intermittent.  Should I be rebasing this branch?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1443424/) on 8038647 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>966449</th>
      <th colspan=4>#23603 803864</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.96%</td>
      <td>84.96%</td>
      <td>+0.00%</td>
      <td>93.60%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>85103</td>
      <td>85202</td>
      <td>+99</td>
      <td>161</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15065</td>
      <td>15077</td>
      <td>+12</td>
      <td>11</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23603/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23603/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23603/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on framework/include/materials/MaterialProperty.h:
```diff
@@ -428,7 +428,13 @@ dataLoad(std::istream & stream, MaterialProperties & v, void * context)
+  const std::size_t old_size = mat_props.size();
+#endif
   loadHelper(stream, mat_props, context);
+  libmesh_assert_equal_to(old_size, mat_props.size());
```
Most typically you will see `mooseAssert` in MOOSE code but if you don't think you could hand-write an assertion error that adds more information than the stack trace, then I'm fine with this

lindsayad commented on framework/include/materials/MaterialPropertyStorage.h:
```diff
@@ -190,45 +192,63 @@ class MaterialPropertyStorage
+  const PropsType & propsOld() const { return *_props_elem_old; }
+  const PropsType & propsOlder() const { return *_props_elem_older; }
+  const MaterialProperties & props(const Elem * elem, unsigned int side) const
   {
-    return *_props_elem;
+    libmesh_assert(props().contains(elem));
+    libmesh_assert(props().find(elem)->second.contains(side));
+    return props().find(elem)->second.find(side)->second;
   }
```
Maybe take these multi-line methods out of the class definition? Can inline them in the header outside the class definition

lindsayad commented on framework/include/materials/MaterialPropertyStorage.h:
```diff
@@ -310,28 +329,47 @@ class MaterialPropertyStorage
+                 unsigned int n_qpoints);
+
+  // You'd think a private mutex would work here, so I'll leave this
+  // in the namespace for when that happens, but CI thinks that can
+  // make us crash, so I'll just initialize this to Threads::spin_mtx
```
How does CI communicate this? I would certainly prefer a class mutex over the global one

roystgnr replied:
Via segfaults, IIRC.  I didn't understand why.  I'd definitely prefer a class mutex too.  Perhaps I used a per-class mutex and a static mutex would behave better?  I'll try that.

lindsayad replied:
Looks like we didn't try the static class mutex?

roystgnr replied:
Not yet.  I'd like to make the upgrade (that's why the reference is here; originally the global mutex was hardcoded elsewhere), but let me try it out in a separate PR?  This one is important enough and urgent enough and invasive enough that I'd rather not get it mixed up with something that might cause intermittent threading race conditions.

lindsayad commented on framework/include/materials/MaterialPropertyStorage.h:
```diff
@@ -310,28 +329,47 @@ class MaterialPropertyStorage
+
+  // Need non-const props from here
+  friend void dataLoad<MaterialPropertyStorage>(std::istream &, MaterialPropertyStorage &, void *);
+  friend void dataStore<MaterialPropertyStorage>(std::ostream &, MaterialPropertyStorage &, void *);
```
Even the store needs non-const?

roystgnr replied:
For now, yeah.  I told everyone I'd try to make `dataStore` const-correct, Logan warned me
```
I tried this
Good luck
This is all I have to say
```
and Cody said "Yeah, I think I tried to fix the const-ness issues there too - good luck!"

... and so when my first attempt seemed to be leading me to a rats nest of compiler errors, I didn't pursue it very far.

loganharbour replied:
Should we put a large comment to this effect at the top of `DataIO.[h,c]`? 😬

loganharbour replied:
@permcody - you should laugh at this too

roystgnr replied:
"// Whoso pulleth out this const-incorrectness of this template and overloads, is rightwise king born of all MOOSE."

lindsayad commented on framework/include/relationshipmanagers/RedistributeProperties.h:
```diff
@@ -0,0 +1,52 @@
+
+  virtual void redistribute() override;
+
+  void addMaterialPropertyStorage(std::vector<std::shared_ptr<MaterialData>> & mat_data,
```
Doxygen

lindsayad commented on framework/include/utils/HashMap.h:
```diff
@@ -32,8 +32,25 @@ class HashMap : public std::unordered_map<Key, T>
+
+  inline iterator find(const Key & k)
+  {
+    libMesh::Threads::spin_mutex::scoped_lock lock(spin_mutex);
```
Yucko. Have we been performing concurrent read-writes to this? Or are we now?

roystgnr replied:
I don't know if we have been in the past, but I got a nasty surprise when I expected the MOOSE "`HashMap` is just a thread-safe `unordered_map`" philosophy to combine cleanly with the C++ "use `find()` instead of `operator[]` when you're not sure an entry exists".

lindsayad commented on framework/include/utils/HashMap.h:
```diff
@@ -32,8 +32,25 @@ class HashMap : public std::unordered_map<Key, T>
+  inline iterator find(const Key & k)
+  {
+    libMesh::Threads::spin_mutex::scoped_lock lock(spin_mutex);
+
```
```suggestion
```

lindsayad commented on framework/include/utils/HashMap.h:
```diff
@@ -32,8 +32,25 @@ class HashMap : public std::unordered_map<Key, T>
+  inline const_iterator find(const Key & k) const
+  {
+    libMesh::Threads::spin_mutex::scoped_lock lock(spin_mutex);
+
```
```suggestion
```

lindsayad commented on framework/src/loops/CacheChangedListsThread.C:
```diff
@@ -27,9 +27,25 @@ CacheChangedListsThread::~CacheChangedListsThread() {}
+  // new children are.  Make sure to cache every such case exactly once.
+  if (elem->refinement_flag() == Elem::JUST_REFINED)
+  {
+    const Elem * parent = elem->parent();
```
```suggestion
    const Elem * const parent = elem->parent();
```

lindsayad commented on framework/src/loops/CacheChangedListsThread.C:
```diff
@@ -27,9 +27,25 @@ CacheChangedListsThread::~CacheChangedListsThread() {}
+    const unsigned int child_num = parent->which_child_am_i(elem);
+
+    bool im_the_lowest_local_child = true;
+    for (unsigned int c = 0; c != child_num; ++c)
```
```suggestion
    for (const auto c : make_range(child_num))
```

lindsayad commented on framework/src/materials/MaterialPropertyStorage.C:
```diff
@@ -417,33 +420,38 @@ MaterialPropertyStorage::initProps(MaterialData & material_data,
+{
+  this->initProps(material_data, *_props_elem, elem, side, n_qpoints);
+  this->initProps(material_data, *_props_elem_old, elem, side, n_qpoints);
+
```
```suggestion
```

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -746,6 +747,53 @@ FEProblemBase::initialSetup()
+    // redistribute() to redistribute properties at the same time
+    // libMesh is redistributing elements.
+    auto add_redistributer =
+        [this](MooseMesh & mesh, const std::string & redistributer_name, bool use_displaced_mesh)
```
```suggestion
        [this](MooseMesh & mesh, const std::string & redistributer_name, const bool use_displaced_mesh)
```

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -746,6 +747,53 @@ FEProblemBase::initialSetup()
+    // properties serialized, so we'll rely on the callback to
+    // redistribute() to redistribute properties at the same time
+    // libMesh is redistributing elements.
+    auto add_redistributer =
```
I would probably vote to make this a class method as `initialSetup` is already very long

lindsayad commented on framework/src/relationshipmanagers/RedistributeProperties.C:
```diff
@@ -0,0 +1,209 @@
+RedistributeProperties::validParams()
+{
+  InputParameters params = RelationshipManager::validParams();
+
```
```suggestion
```

lindsayad commented on framework/src/relationshipmanagers/RedistributeProperties.C:
```diff
@@ -0,0 +1,209 @@
+
+    if (mat_prop_store->hasStatefulProperties())
+    {
+      MaterialData & my_mat_data = *((*mat_data)[/*_tid*/ 0]); // Not threaded
```
Can we assert then that we are `!in_threads`? I get nervous after my coupling functor was not thread safe (I know that you are only adding this to the mesh but still like to assert)

roystgnr replied:
Kinda implied by the API that's being called, IMHO (any threading ought to be done inside redistribute() implementations, and the fact that it's a function applied to a whole mesh with no range-of-elements should be enough to deduce that), but I can't actually bring myself to say the phrase "this code shouldn't have so many assertions", so I'll add it.

lindsayad commented on framework/src/relationshipmanagers/RedistributeProperties.C:
```diff
@@ -0,0 +1,209 @@
+          mat_prop_store->_props_elem_old.get(),
+          mat_prop_store->_props_elem_older.get()};
+
+      for (auto * props_map_ptr : props_maps)
```
```suggestion
      for (auto * const props_map_ptr : props_maps)
```

lindsayad commented on framework/src/relationshipmanagers/RedistributeProperties.C:
```diff
@@ -0,0 +1,209 @@
+              {
+                libmesh_assert(!prop_vals.empty());
+
+                for (PropertyValue * prop : prop_vals)
```
```suggestion
                for (PropertyValue * const prop : prop_vals)
```

lindsayad commented on framework/src/relationshipmanagers/RedistributeProperties.C:
```diff
@@ -0,0 +1,209 @@
+            elems_being_migrated.insert(elem);
+        }
+
+        auto recv_functor = [&, mat_prop_store_ptr = mat_prop_store](
```
I've never seen this before. This performs a copy I take it?

roystgnr replied:
Yup.  Fortunately a pointer copy.

There was a big rat hole behind this: https://stackoverflow.com/a/46115028

lindsayad commented on test/tests/materials/stateful_prop/spatial_adaptivity_test.i:
```diff
@@ -4,11 +4,6 @@
   nx = 2
   ny = 2
   uniform_refine = 3
-  # This option is necessary if you have uniform refinement + stateful material properties + adaptivity
-  skip_partitioning = true
-  # stateful material properties + adaptivity are not yet compatible
-  # with distributed meshes
-  parallel_type = replicated
```
love it

roystgnr committed:
Better assert message

roystgnr committed:
Move inline functions out of class definition

roystgnr committed:
Doxygen for RedistributeProperties methods

roystgnr committed:
Whitespace removal

In my defense, my new methods were just emulating the existing methods.

roystgnr committed:
constness, use make_range

roystgnr committed:
Factor out FEProblemBase::addAnyRedistributers()

roystgnr committed:
assert(!in_threads)

We're definitely not in threads, but we *definitely* shouldn't
discourage developers who want more assertions.

roystgnr commented:
Rebased on next to clean up the conflict; this ought to be ready to re-review now.

roystgnr commented:
We broke (well, added better encapsulation in) next again?  Those app failures are totally not my fault.

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1464757/) on fcba616 : invalidated by @lindsayad

moosebuild commented:
Job [Apptainer build apps](https://civet.inl.gov/job/1464737/) on fcba616 : invalidated by @lindsayad

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/1464760/) on fcba616 : invalidated by @lindsayad

GiudGiud commented:
should add a newsletter entry but i m in favor of follow up PRs for this

moosebuild commented:
Job [App documentation](https://civet.inl.gov/job/1464757/) on fcba616 : invalidated by @roystgnr

&quot;Connection reset by peer&quot; while trying to build Grizzly docs

GiudGiud commented on framework/src/relationshipmanagers/RedistributeProperties.C:
```diff
@@ -0,0 +1,212 @@
+        for (auto & [elem, elem_props_map] : props_map)
+        {
+          // There better be an element here
+          libmesh_assert(elem);
```
I prefer mooseAssert for future reference. No message is not super friendly to the next person hitting this

GiudGiud replied:
I dont think it's because of a policy / difficulty to add a string to mooseAssert that we have less asserts than libmesh does. 

My guess would be because expertise about what could go wrong is much stronger within the libmesh devs than the MOOSE ones.

roystgnr committed:
Newsletter entry

roystgnr committed:
libmesh_assert -> mooseAssert
